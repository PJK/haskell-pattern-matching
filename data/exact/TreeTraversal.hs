module TreeTraversal where

data Tree a = Fork a (Tree a) (Tree a) | Nil

func :: Tree a -> String
func Nil = "Empty tree"
func (Fork _ Nil Nil) = "Leaf"
-- At this point, we need correct propagation of value abstraction to handle this precisely, i.e.
-- deduce that that the last parameter can only be Fork
func (Fork _ Nil _) = "Right subtree only"
func (Fork _ _ Nil) = "Left subtree only"
-- Fork _ F F missing
