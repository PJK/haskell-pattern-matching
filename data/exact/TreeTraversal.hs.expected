{
    "tag": "AnalysisError",
    "contents": {
        "tag": "ProcessError",
        "contents": {
            "tag": "VariableNotBound",
            "contents": "Type lookup for variable fresh7 failed (searched in fromList [(\"fresh0\",TypeApplication (TypeConstructor \"Tree\") (VariableType \"a\")),(\"fresh4\",VariableType \"a\"),(\"fresh5\",TypeApplication (TypeConstructor \"Tree\") (VariableType \"a\")),(\"fresh6\",TypeApplication (TypeConstructor \"Tree\") (VariableType \"a\"))])"
        }
    }
}